lib/smack/SmackInstGenerator.cpp

```diff
@@ -144,6 +144,37 @@ void SmackInstGenerator::processInstruction(llvm::Instruction& inst) {
   nextInst++;
 }
 
+void SmackInstGenerator::processIntrinsicCall(llvm::IntrinsicInst* ii) {
+  static auto f16Cast = [&ii, this](bool f16UpCast) {
+    if (SmackOptions::BitPrecise && SmackOptions::FloatEnabled) {
+      std::string funcName;
+      auto tyID = f16UpCast?
+        ii->getFunctionType()->getReturnType()->getTypeID():
+        ii->getArgOperand(0)->getType()->getTypeID();
+      if (tyID == llvm::Type::FloatTyID)
+        funcName = f16UpCast? "__SMACK_convert_from_fp16_f32":
+          "__SMACK_convert_to_fp16_f32";
+      else if (tyID == llvm::Type::DoubleTyID)
+        funcName = f16UpCast? "__SMACK_convert_from_fp16_f64":
+          "__SMACK_convert_to_fp16_f64";
+      return Stmt::call(funcName, {rep->expr(ii->getArgOperand(0))},
+            {naming->get(*ii)});
+    } else
+      return rep->call(ii->getCalledFunction(), *ii);
+  };
+  static std::map<llvm::Intrinsic::ID, std::function<const Stmt*()>> stmtMap {
+    {llvm::Intrinsic::convert_from_fp16,
+      []() { return f16Cast(true); }},
+    {llvm::Intrinsic::convert_to_fp16,
+      []() { return f16Cast(false); }}
+  };
+  auto it = stmtMap.find(ii->getIntrinsicID());
```

Comment:
I don't understand the purpose of `stmtMap`. It is generated dynamically every time `processIntrinsicCall` is called. So how come you are putting the generated statements in this map, instead of emitting them directly?

Reply:
Actually this map is not named properly. It should be something like `stmtGenMap` which means that it maps an intrinsic ID to a procedure that generates Boogie statements. Such a map is generated only once as the `static` keyword in its definition suggests. The Boogie statement is generated dynamically via application to the mapped procedure. I guess the purpose of this map is because one-to-one mapping from intrinsic IDs to generated statements is not possible since the intrinsic IDs are overloaded. However, the one-to-one mapping from intrinsic IDs to the statement generation procedure is possible.